### PR TITLE
[feat](nereids) disable join reorder if any table row count is not available

### PR DESCRIPTION
**Pull request by englefly**
## Proposed changes
 join reorder
if a table is not analyzed, and its row count is not fully reported, optimizer will disable join reorder.
fully report means after a table is newly created or truncated, all tablets row count are reported.

Issue Number: close #xxx

### CODE REVIEW

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java [hunk 1]
```diff
@@ -205,6 +210,22 @@ public Map<String, ColumnStatistic> getTotalColumnStatisticMap() {
...
 
+    /**
+     * disable join reorder if any table row count is not available.
+     */
+    public static void disableJoinReorderIfTableRowCountNotAvailable(
+            List<LogicalOlapScan> scans, CascadesContext context) {
+        StatsCalculator calculator = new StatsCalculator(context);
+        for (LogicalOlapScan scan : scans) {
+            double rowCount = calculator.getOlapTableRowCount(scan);
+            if (rowCount == -1 && ConnectContext.get() != null) {
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java:221]: better to add comments here to notice what is the "-1"'s indication.

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java [hunk 2]
```diff
@@ -288,6 +296,18 @@ private Plan planWithoutLock(
         return physicalPlan;
     }
 
+    private List<LogicalOlapScan> getAllOlapScans(Plan plan) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java:299]: could replace by `plan.collect(LogicalOlapScan.class::isInstance)`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java [hunk 3]
```diff
@@ -205,6 +210,23 @@ public Map<String, ColumnStatistic> getTotalColumnStatisticMap() {
...
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java:221]: add comment here to explain magic number `-1`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java (hunk 1 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java:221]: the meaning of -1 is explained in comments for getOlapTableRowCount()

File: fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java (hunk 2 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java:299]: the meaning of -1 is explained in comments for getOlapTableRowCount()

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run p0

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run cloud_p0

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java [hunk 4]
```diff
@@ -256,6 +259,17 @@ private Plan planWithoutLock(
...
         }
 
+        // if we cannot get table row count, skip join reorder
+        // except:
+        //   1. user set leading hint
+        //   2. ut test. In ut test, FeConstants.enableInternalSchemaDb is false or FeConstants.runningUnitTest is true
+        if (FeConstants.enableInternalSchemaDb && !FeConstants.runningUnitTest
+                && !cascadesContext.isLeadingDisableJoinReorder()) {
+            List<LogicalOlapScan> scans = cascadesContext.getRewritePlan()
+                    .collectToList(LogicalOlapScan.class::isInstance);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java:269]: only need olapscan, how about other scan?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java (hunk 3 above)
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java:221]: -1 is a magic number could we add a static variable `UNKNOWN_ROW_COUNT` to static utils and replace all magic number `-1` by the static variable

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java [hunk 5]
```diff
@@ -205,6 +210,23 @@ public Map<String, ColumnStatistic> getTotalColumnStatisticMap() {
...
+                LOG.info("disable join reorder since row count not available: "
+                        + scan.getTable().getNameWithFullQualifiers());
+                ConnectContext.get().getSessionVariable().setDisableJoinReorder(true);
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java:224]: This sets session variable directly, should use disableNereidsJoinReorderOnce?

Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java:224]: this is statement level. it has no impact on other queries.

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java (hunk 3 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java:221]: this is the return value of OlapTable.getRowCountForIndex().
explained the "-1" at the spec of OlapTable.getRowCountForIndex()

File: fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java (hunk 4 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/NereidsPlanner.java:269]: this function depends on OlapTable.getRowCountForIndex().

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run cloud_p0

**Comment by englefly** [comment]
run buildall

File: regression-test/suites/nereids_rules_p0/push_down_filter/extract_from_disjunction_in_join.groovy [hunk 6]
```diff
@@ -20,7 +20,11 @@ suite("extract_from_disjunction_in_join") {
     sql "SET enable_fallback_to_original_planner=false"
     sql "set ignore_shape_nodes='PhysicalDistribute,PhysicalProject'"
     sql "set disable_nereids_rules=PRUNE_EMPTY_PARTITION"
-    sql "set runtime_filter_mode=OFF"
+    sql """
+        set runtime_filter_mode=OFF;
+        set disable_join_reorder=true;
+        set disable_join_reorder=true;
```
Comment (xzj7019) [inline regression-test/suites/nereids_rules_p0/push_down_filter/extract_from_disjunction_in_join.groovy:26]: dup

**Comment by englefly** [comment]
run cloud_p0

**Review by starocean999 — APPROVED** [review]

**Review by xzj7019 — APPROVED** [review]

**Pull request merged**